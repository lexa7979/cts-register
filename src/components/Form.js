/**
 * MIT License
 *
 * Copyright (c) 2019 <alexander.urban@cygni.se>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import assert from "assert";

import React from "react";
import PropTypes from "prop-types";

import "./Form.scss";

const propTypes = {
	formClass:    PropTypes.string,
	fields:       PropTypes.object.isRequired,
	handleSubmit: PropTypes.func.isRequired,
};

const defaultProps = {};

/**
 * Component which contains an input form.
 *
 * @property Component.fields
 *		The needed input fields must be given as an object-property "fields".
 *		The following input types are supported:
 *		- "input" with optional settings "label", "value", "placeholder", e.g.
 *				fullname: { type: "input", label: "What's your name?" }
 *		- "textarea" with optional settings "label", "value", "placeholder", "cols", "rows", e.g.
 *	 			comments: { type: "textarea", placeholder: "Any comments?", rows: 3 }
 *		- "select"
 *		- "checkbox"
 *		- "radio"
 *		- "hidden"
 *		- "submit"
 */
export class Form extends React.Component {
	// eslint-disable-next-line require-jsdoc
	constructor( props ) {
		super( props );

		if (
			this.props.fields === null
			|| typeof this.props.fields !== "object"
			|| typeof this.props.handleSubmit !== "function"
		) {
			return;
		}

		this.state = {
			inputValues: {},
		};

		const { fields } = this.props;
		this.fieldList = [];
		for ( const name in fields ) {
			if ( typeof name === "string" && name !== "" ) {
				const { type = "input", value = "" } = fields[name];
				if (
					typeof this[`${type}FieldGenerator`] !== "function"
					&& typeof Form[`${type}FieldGenerator`] !== "function"
				) {
					throw new Error( `Unsupported field type "${type}" (${name})` );
				}
				this.fieldList[name] = { ...fields[name], name, type };
				this.state.inputValues[name] = value;
			}
		}

		this.handleInputChange = this.handleInputChange.bind( this );
		this.onSubmit = this.onSubmit.bind( this );
	}

	/**
	 * Wraps the given input component into a new div-container.
	 * A label-element will be added if requested.
	 *
	 * @param	{string} 	inputName
	 * @param	{null|string} 	label
	 * @param	{null|string}	cssClasses
	 */
	static labelGenerator( inputName, label, cssClasses ) {
		assert( typeof inputName === "string" && inputName !== "", "Invalid argument" );
		assert( label == null || ( typeof label === "string" && label !== "" ), "Invalid argument" );
		assert( cssClasses == null || typeof cssClasses === "string", "Invliad argument" );

		if ( label ) {
			return <label key={`${inputName}-label`}
				className={`label ${cssClasses || ""}`}
				htmlFor={inputName}
			>{label}</label>;
		}

		return <div key={`${inputName}-nolabel`}
			className={`nolabel ${cssClasses || ""}`}
		></div>;
	}

	/**
	 * Delivers the components for a new text field.
	 *
	 * @param	{null|string}	data.label
	 * @param	{string}	data.name
	 * @param	{null|string}	data.value
	 * @param	{null|string}	data.placeholder
	 */
	inputFieldGenerator( data ) {
		assert( data !== null && typeof data === "object", "Invalid argument" );

		return <input key={data.name}
			name={data.name}
			type="text"
			value={this.state.inputValues[data.name]}
			placeholder={data.placeholder || null}
			className="field input"
			onChange={this.handleInputChange}
		/>;
	}

	/**
	 * Delivers the components for a new multiline text field.
	 *
	 * @param	{null|string}	data.label
	 * @param	{string}	data.name
	 * @param	{null|string}	data.value
	 * @param	{null|string}	data.placeholder
	 * @param	{null|number}	data.cols
	 * @param	{null|number}	data.rows
	 */
	textareaFieldGenerator( data ) {
		assert( data !== null && typeof data === "object", "Invalid argument" );

		return <textarea key={data.name}
			name={data.name}
			value={this.state.inputValues[data.name] || ""}
			placeholder={data.placeholder || null}
			cols={data.cols || null}
			rows={data.rows || null}
			className="field textarea"
			onChange={this.handleInputChange}
		/>;
	}

	/**
	 * Delivers the components for a new select box.
	 *
	 * @param	{object}	data
	 * /
	selectFieldGenerator( data ) {
		assert( data !== null && typeof data === "object", "Invalid argument" );
	}

	/**
	 * Delivers the components for a new checkbox field.
	 *
	 * @param {*} data
	 * /
	checkboxFieldGenerator( data ) {
		assert( data !== null && typeof data === "object", "Invalid argument" );
	}

	/**
	 * Delivers the components for a new group of radio buttons.
	 *
	 * @param	{*}	data
	 */
	radioFieldGenerator( data ) {
		assert( data !== null && typeof data === "object", "Invalid argument" );

		const buttonList = [];
		for ( let index = 0; index < data.options.length; index++ ) {
			const id = `${data.name}-radio-${index + 1}`;
			buttonList.push( <div key={id}>
				<input
					type="radio"
					id={id}
					name={data.name}
					value={data.options[index]}
					onChange={this.handleInputChange}
				/>
				<label htmlFor={id}>{data.options[index]}</label>
			</div> );
		}

		return <div key={data.name} className="field radio">{buttonList}</div>;
	}

	/**
	 * Delivers the components for a new hidden field.
	 *
	 * @param	{*}	data
	 * /
	hiddenFieldGenerator( data ) {
		assert( data !== null && typeof data === "object", "Invalid argument" );

		return <input type="hidden" key={data.name} name={data.name} value={data.value} />;
	}

	/**
	 * Delivers the component for a new submit button.
	 *
	 * @param	{*}	data
	 */
	static submitFieldGenerator( data ) {
		assert( data !== null && typeof data === "object", "Invalid argument" );

		return <input key={data.name || "submit"}
			type="submit"
			name={data.name || null}
			value={ data.value || "Submit" }
			className="field submit"
		/>;
	}

	/**
	 * Handles the change event of any form field
	 *
	 * @param	{object}	event
	 *		Description of the event to handle
	 */
	handleInputChange( event ) {
		const { name } = event.target;
		const value = event.target.type === "checkbox" ? event.target.checked : event.target.value;

		this.setState( { inputValues: { ...this.state.inputValues, [name]: value } } );
	}

	/**
	 *
	 */
	onSubmit( event ) {
		event.preventDefault();

		if ( typeof this.props.handleSubmit === "function" ) {
			this.props.handleSubmit( this.state.inputValues );
		}
	}

	// eslint-disable-next-line require-jsdoc
	render() {
		if ( this.fieldList === null || typeof this.fieldList !== "object" ) {
			return <div className="empty-form"></div>;
		}

		const formFields = [];
		const submitFields = [];
		for ( const name in this.fieldList ) {	// eslint-disable-line guard-for-in
			const data = this.fieldList[name];
			if ( data.type === "submit" ) {
				submitFields.push( Form.submitFieldGenerator( data ) );
			} else if ( typeof this[`${data.type}FieldGenerator`] === "function" ) {
				formFields.push( Form.labelGenerator( data.name, data.label, data.type ) );
				formFields.push( this[`${data.type}FieldGenerator`]( data ) );
			} else {
				formFields.push( Form.labelGenerator( data.name, data.label, data.type ) );
				formFields.push( Form[`${data.type}FieldGenerator`]( data ) );
			}
		}

		if ( submitFields.length === 0 ) {
			submitFields.push( Form.submitFieldGenerator( { value: "Submit" } ) );
		}

		return <form
			className={this.props.formClass ? `form ${this.props.formClass}` : "form"}
			onSubmit={this.onSubmit}
		>
			<div className="fields">
				{formFields}
			</div>
			<div className="final">
				{submitFields}
			</div>
		</form>;
	}
}

Form.propTypes = propTypes;
Form.defaultProps = defaultProps;

export default Form;
