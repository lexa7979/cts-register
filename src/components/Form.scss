/**
 * MIT License
 *
 * Copyright (c) 2019 <alexander.urban@cygni.se>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

@import "../mixins";

.form {
	@include flex( column, $horizontal: center );

	.fields {
		width: 100%;

		display: grid;
		grid-template-columns: 50% 50%;
		grid-row-gap: 10px;

		.label {
			box-sizing: border-box;
			padding-right: 5px;

			justify-self: end;
			align-self: center;
			text-align: right;

			font-size: 0.8em;
			font-weight: bold;
			line-height: 1.5em;
		}

		.field {
			box-sizing: border-box;
			width: 100%;

			justify-self: start;
			align-self: center;
		}

		.field.input {
			box-sizing: border-box;
			padding: 8px 4px;

			border-radius: 3px;
			border: 2px solid #9c4;

			font-size: 1em;
		}

		.field.radio {
			box-sizing: border-box;
			padding: 8px 0;

			@include flex( row, wrap, space-around );

			& > div > input, & > div > label {
				cursor: pointer;
			}
		}

		.message {
			grid-column: 1 / 3;

			box-sizing: border-box;
			padding: 0 0 5px 0;

			font-size: 0.7em;
			font-weight: bold;
			text-align: center;
			color: red;
		}
	}

	.final {
		align-self: flex-start;
		box-sizing: border-box;
		width: 100%;
		padding: 5px 0;

		@include flex( row, nowrap, flex-end );

		input[type=submit] {
			padding: 7px 12px;
			background: #dfd;

			border-radius: 10px;
			border: 2px solid #9c4;

			color: #6a4;
			font-size: 1em;
			font-weight: bold;
		}

		input[type=submit]:hover {
			border: 2px solid #060;
			color: #060;
			cursor: pointer;
		}

		input[type=submit]:disabled {
			background: white;
			border: 2px solid #777;
			color: #777;
			cursor: default;
		}
	}
}
