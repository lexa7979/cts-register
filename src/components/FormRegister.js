/**
 * Copyright (c) 2019 <alexander.urban@cygni.se>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import React from "react";

import Form from "./Form";

/**
 * Component which will be used by users to register for the event.
 */
export class FormRegister extends React.Component {
	// eslint-disable-next-line require-jsdoc
	constructor( props ) {
		super( props );

		this.formFields = {
			firstname: { type: "input", label: "Firstname:", placeholder: "Type in your firstname?" },
		};

		// eslint-disable-next-line no-console
		this.handleSubmit = inputData => console.log( "Form was submitted", inputData );
	}

	// eslint-disable-next-line require-jsdoc
	render() {
		return <Form formClass="register" fields={this.formFields} handleSubmit={this.handleSubmit} />;
	}
}
