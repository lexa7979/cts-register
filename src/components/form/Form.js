/**
 * MIT License
 *
 * Copyright (c) 2019 <alexander.urban@cygni.se>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

import assert from "assert";

import React from "react";
import PropTypes from "prop-types";

import FormTypes from "./types";

import "./Form.scss";

const propTypes = {
	fields:       PropTypes.object.isRequired,
	handleSubmit: PropTypes.func.isRequired,
	formClass:    PropTypes.string,
	validateData: PropTypes.func,
};

const defaultProps = {};

/**
 * Component which contains an input form
 */
export class Form extends React.Component {
	/**
	 * Initialising component
	 */
	constructor( props ) {
		super( props );

		this.state = {
			inputValues: {},
			touched:     {},
			messages:    {},
			actions:     [],
		};

		this.fieldList = [];
		for ( const name in this.props.fields ) {
			if ( typeof name === "string" && name !== "" ) {
				const { type = "input", value = "" } = this.props.fields[name];
				assert(
					FormTypes[type] != null
					&& typeof FormTypes[type] === "object"
					&& typeof FormTypes[type].generateField === "function",
					`Unsupported field type "${type}" (${name})`
				);
				this.fieldList[name] = {
					...this.props.fields[name],
					name,
					type,
				};
				this.state.inputValues[name] = value;
			}
		}

		this.validation = {
			progressing: false,
			requested:   false,
		};

		this.handleInputChange = this.handleInputChange.bind( this );
		this.onSubmit = this.onSubmit.bind( this );
	}

	/**
	 * Actions to take when component is ready
	 */
	componentDidMount() {
		this.runValidation();
	}

	/**
	 * Validates the form inputs if props.validateData is given
	 *
	 * If another (asynchronous) validation is already in progress,
	 * nothing will be done, now. Still, it's memorised that
	 * another validation shall be run just after the active one
	 * has finished.
	 */
	runValidation() {
		if ( typeof this.props.validateData !== "function" ) {
			return Promise.resolve();
		}

		if ( this.validation.progressing ) {
			this.validation.requested = true;
			return Promise.resolve();
		}
		this.validation.progressing = true;

		return Promise.resolve( this.props.validateData( this.state.inputValues ) )
			.then( result => {
				console.log( result );

				if (
					result != null && typeof result === "object"
					&& Array.isArray( result.actions )
					&& result.messages != null && typeof result.messages === "object"
				) {
					this.setState( { actions: result.actions, messages: result.messages } );
				}

				this.validation.progressing = false;
				if ( this.validation.requested ) {
					this.validation.requested = false;
					this.runValidation();
				}
			} );
	}

	/**
	 * Wraps the given input component into a new div-container.
	 * A label-element will be added if requested.
	 *
	 * @param	{string} 	inputName
	 * @param	{null|string} 	label
	 * @param	{null|string}	cssClasses
	 */
	static labelGenerator( inputName, label, cssClasses ) {
		assert( typeof inputName === "string" && inputName !== "", "Invalid argument" );
		assert( label == null || ( typeof label === "string" && label !== "" ), "Invalid argument" );
		assert( cssClasses == null || typeof cssClasses === "string", "Invliad argument" );

		if ( label ) {
			return <label key={`${inputName}-label`}
				className={`label ${cssClasses || ""}`}
				htmlFor={inputName}
			>{label}</label>;
		}

		return <div key={`${inputName}-nolabel`}
			className={`nolabel ${cssClasses || ""}`}
		></div>;
	}

	/**
	 * If there is a message for the given input field,
	 * a matching component to show this text is composed.
	 *
	 * @param	{string}	inputName
	 */
	messageGenerator( inputName ) {
		assert( typeof inputName === "string" && inputName !== "", "Invalid argument" );

		if ( !this.state.touched[inputName] || !this.state.messages[inputName] ) {
			return null;
		}

		return <div	key={`${inputName}-message`} className="message">
			{this.state.messages[inputName]}
		</div>;
	}

	/**
	 * Handles the change event of any form field
	 *
	 * @param	{object}	event
	 *		Description of the event to handle
	 */
	handleInputChange( event ) {
		const { name } = event.target;
		const value = event.target.type === "checkbox" ? event.target.checked : event.target.value;

		const stateUpdate = {
			inputValues: {
				...this.state.inputValues,
				[name]: value,
			},
		};

		if ( this.state.touched[name] !== true ) {
			stateUpdate.touched = {
				...this.state.touched,
				[name]: true,
			};
		}

		this.setState( stateUpdate, () => {
			this.runValidation();
		} );
	}

	/**
	 * Handles events whenever a user clicks on one of the submit buttons
	 */
	onSubmit( event ) {
		event.preventDefault();

		if ( typeof this.props.handleSubmit === "function" ) {
			this.props.handleSubmit( this.state.inputValues );
		}
	}

	/**
	 * Composing output
	 */
	render() {
		if ( this.fieldList === null || typeof this.fieldList !== "object" ) {
			return <div className="empty-form"></div>;
		}

		const formFields = [];
		const submitFields = [];
		for ( const name in this.fieldList ) {	// eslint-disable-line guard-for-in
			const data = this.fieldList[name];
			if ( data.type === "submit" ) {
				submitFields.push( FormTypes.submit.generateField.call( this, data ) );
			} else if ( typeof FormTypes[data.type] === "object" ) {
				formFields.push( Form.labelGenerator( data.name, data.label, data.type, data.required ) );
				formFields.push( FormTypes[data.type].generateField.call( this, data ) );
			}

			const message = this.messageGenerator( data.name );
			if ( message ) {
				formFields.push( message );
			}
		}

		if ( submitFields.length === 0 ) {
			submitFields.push( FormTypes.submit.generateField.call( this, { value: "Submit" } ) );
		}

		return <form
			className={this.props.formClass ? `form ${this.props.formClass}` : "form"}
			onSubmit={this.onSubmit}
		>
			<div className="fields">
				{formFields}
			</div>
			<div className="final">
				{submitFields}
			</div>
		</form>;
	}
}

Form.propTypes = propTypes;
Form.defaultProps = defaultProps;

export default Form;
