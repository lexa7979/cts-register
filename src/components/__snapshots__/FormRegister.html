
<!DOCTYPE html>
<html style="height: 100%;">
  <head>
  </head>
  <body style="display: flex; flex-flow: column nowrap; justify-content: center; align-items: center; height: 100%;">
    <h2>
      Component FormRegister - when rendering with no properties - delivers expected result  (-> check snapshot, too)
    </h2>
    <form class="form register">
      <div class="fields">
        <label class="label input"
               for="firstname"
        >
          Firstname:
        </label>
        <input type="text"
               name="firstname"
               value
               placeholder="Type in your firstname."
               class="field input"
        >
        <label class="label input"
               for="lastname"
        >
          Lastname:
        </label>
        <input type="text"
               name="lastname"
               value
               placeholder="Type in your lastname."
               class="field input"
        >
        <label class="label radio"
               for="attending"
        >
          Will you attend the conference?
        </label>
        <div class="field radio">
          <div>
            <input type="radio"
                   id="attending-radio-1"
                   name="attending"
                   value="yes"
            >
            <label for="attending-radio-1">
              yes
            </label>
          </div>
          <div>
            <input type="radio"
                   id="attending-radio-2"
                   name="attending"
                   value="no"
            >
            <label for="attending-radio-2">
              no
            </label>
          </div>
          <div>
            <input type="radio"
                   id="attending-radio-3"
                   name="attending"
                   value="maybe"
            >
            <label for="attending-radio-3">
              maybe
            </label>
          </div>
        </div>
      </div>
      <div class="final">
        <input type="submit"
               name="submit"
               value="Submit"
               class="field submit"
        >
      </div>
    </form>
    <br>
    <br>
    &lt;FormRegister /&gt;
  </body>
</html>
